cache/sharded_cache.cc

```diff
@@ -59,8 +60,15 @@ Status ShardedCache::Insert(const Slice& key, void* value, size_t charge,
                             DeleterFn deleter, Handle** handle,
                             Priority priority) {
   uint32_t hash = HashSlice(key);
-  return GetShard(Shard(hash))
-      ->Insert(key, hash, value, charge, deleter, handle, priority);
+  Status s = GetShard(Shard(hash))
+                 ->Insert(key, hash, value, charge, deleter, handle, priority);
+  if (s.ok() && cache_res_mgr_) {
+    // Insert may cause the cache entry eviction if the cache is full. So we
+    // directly call the reservation manager to update the total memory used
+    // in the cache.
+    cache_res_mgr_->UpdateCacheReservation(GetUsage()).PermitUncheckedError();
+  }
+  return s;
```

Comment:
We might want to add some validation to ensure that such a "cache reservation aware" cache is configured as a blob cache if and only if charging is enabled for `CacheEntryRole::kBlobCache`.

Reply:
An extra relevant idea: should we add some kind of early option validation in block based table factory to return InvalidArgument when CacheEntryRole::kBlobCache is enabled AND blob cache == block cache

Reply:
>  if and only if charging is enabled for CacheEntryRole::kBlobCache.

If we do open up API usage like https://github.com/facebook/rocksdb/pull/10321/files#diff-5c4ced6afb6a90e27fec18ab03b2cd89e8f99db87791b4ecc6fa2694284d50c0R2921-R2925 to users, update to BlockBasedTableOptions::cache_usage_options API is needed. 

Also, should we consider early option validation like https://github.com/facebook/rocksdb/blob/7.4.fb/table/block_based/block_based_table_factory.cc#L709-L714 but for blob cache when no blob cache is specified or blob cache = block cache when CacheEntryRole::kBlobCache is enabled?

Reply:
> Instead of adding the cache reservation manager directly to `ShardedCache`, we could consider moving this logic a separate class which also implements the `Cache` interface, wraps another `Cache`, takes care of the cache reservations and forwards calls to the wrapped cache. This would be a more decoupled design, and would also make the charging logic work with any custom cache implementations.

Where is this cache wrapper created? It looks like the user can create it and pass it to `options.blob_cache`?

Reply:
> Where is this cache wrapper created? It looks like the user can create it and pass it to `options.blob_cache`? In this case, it's hard to add some validation. The way @hx235 said above is feasible, even if the `options.blob_cache` settings here are wrong, we are able to detect anomalies from option validation in `block_based_table_factory` and throw an error.

Yes, it would be created by the application and assigned to `options.blob_cache`. During validation, we could use e.g. `dynamic_cast` to ensure the application configured the right type of cache.

Reply:
Sorry, I'm kind of having second thoughts re: having the application create the wrapper. I'm wondering if it would be better if we created it e.g. in `BlobSource` ? I feel it would be more user-friendly and would hide this implementation detail. The downside of doing so would be that the book-keeping would be bypassed if the application manipulates the underlying cache directly but that's not something they should be doing anyways...

Reply:
No problem, I also wanted to create a wrapper in the blob source at first. But if we create the cache wrapper in the blob source, how to ensure that rocksdb will call the cache wrapper's release or erase.